File: 0103-binary-tree-zigzag-level-order-traversal/solution.cpp

class Solution {
public:
    int level(TreeNode* root) {
        if (root == nullptr)
            return 0;
        return 1 + max(level(root->left), level(root->right));
    }
    void displaypre(TreeNode* root, int curr, int max,
                    vector<int>& w) // left ORDER displaypre
    {
        if (root == nullptr)
            return;
        if (curr == max) {
            w.push_back(root->val);
            return;
        }
        displaypre(root->left, curr + 1, max, w);
        displaypre(root->right, curr + 1, max, w);
    }
    void displayprerev(TreeNode* root, int curr, int max,
                       vector<int>& w) // left ORDER displaypre
    {
        if (root == nullptr)
            return;
        if (curr == max) {
            w.push_back(root->val);
            return;
        }
        displayprerev(root->right, curr + 1, max, w);
        displayprerev(root->left, curr + 1, max, w);
    }
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> v;
        for (int i = 1; i <= level(root); i++) {
            vector<int> w;
            if (i % 2 == 0) {
                displayprerev(root, 1, i, w);
            } else {
                displaypre(root, 1, i, w);
            }
            v.push_back(w);
            cout << endl;
        }
        return v;
    }
};

File: 0103-binary-tree-zigzag-level-order-traversal/solution.java
import java.util.*;

class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null)
            return result; // edge case

        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        
        // Start check at 0. Level 0 (root) is even, so it won't reverse.
        int check = 0; 
        
        while (!q.isEmpty()) {
            int size = q.size(); // current level size
            List<Integer> level = new ArrayList<>();

            for (int i = 0; i < size; i++) {
                TreeNode curr = q.poll();
                level.add(curr.val);

                if (curr.left != null)
                    q.offer(curr.left);
                if (curr.right != null)
                    q.offer(curr.right);
            }
            
            // If check % 2 == 1 (odd levels like Level 1, 3, etc.), reverse it
            if (check % 2 != 0) {
                Collections.reverse(level);
            } 
            
            // You don't need a dummy copy here anymore!
            // 'level' is created fresh inside the while-loop on line 16, 
            // so adding it directly to 'result' is perfectly safe.
            result.add(level);
            
            check += 1; // Move to the next level indicator
        }
        return result;
    }
}

